// @flathockey/shared/src/tuning/movement.defaults.ts
import type { MovementStepConfig } from '../sim/movementStep';

export type MovementTuning = MovementStepConfig & {
  __version?: number;
};

export const MOVEMENT_DEFAULTS: MovementTuning = {
  maxSpeed: 342.5,
  maxSpeedNoPuck: 342.5,
  maxSpeedWithPuck: 342.5,
  accel: 1681.36,
  sprintAccel: 2800,
  dragMove: 2.75,
  dragIdle: 0.96909,
  brakeDrag: 11,
  lateralGrip: 1.13636,
  reverseBrake: 0,
  brakeCurve: 0.7545,
  overspeedDamping: 1,
  gripCurve: 0,
  hasPuck: false,
  sprintMulNoPuck: 1,
  sprintMulWithPuck: 1,
  sprintMinStamina: 0.1,
  staminaDrain: 0.38,
  staminaRegen: 0.23,
  staminaDrainMulWithPuck: 1.2,
  headingModeEnabled: true,
  maxTurnRateLowSpeed: 10.5,
  maxTurnRateHighSpeed: 2.6,
  lateralDamping: 0.98,
  steeringEnabled: false,
  steerStrength: 6,
  brakeDecel: 20,
  turnAssist: 0,
  driftAssist: 0,
  regimesEnabled: true,
  speedSplit: 0.55,
  splitBlendWidth: 0.12,
  accel_lo: 1400,
  dragMove_lo: 3.025,
  dragIdle_lo: 0.96909,
  lateralGrip_lo: 1.36363,
  brakeCurve_lo: 0.7922,
  overspeedDamping_lo: 1,
  accel_hi: 1429.156,
  dragMove_hi: 2.3375,
  dragIdle_hi: 0.920636,
  lateralGrip_hi: 0.9659,
  brakeCurve_hi: 0.71678,
  overspeedDamping_hi: 1,
  inputDeadzone: 0,
  inputExponent: 1,

  // stick (v1)
  stickOffsetX: 22,
  stickOffsetY: 0,
  stickLength: 30,
  stickTipRadius: 12,
  stickVisualLag: 0.2,
  stickVisualLagMaxDeg: 18,
  drawStickTarget: false,
  drawStickHitbox: false,

  // puck core
  puckRadius: 8,
  puckMaxSpeed: 560,
  puckLinearDamping: 3.2,
  puckRestitution: 0.18,
  puckSurfaceDrag: 0.04,

  // puck pickup / held
  puckPickupRadius: 22,
  puckPickupMaxSpeed: 220,
  puckPickupMaxRelativeSpeed: 180,
  puckMagnetRadius: 30,
  puckMagnetStrength: 90,
  puckMagnetMaxForce: 120,
  puckHoldSpringK: 28,
  puckHoldDampingC: 10,
  puckHoldMaxError: 46,
  puckPickupCooldownMs: 280,

  // puck shot
  puckShotBaseImpulse: 240,
  puckShotChargeRate: 1.8,
  puckShotChargeMult: 300,
  puckShotMaxImpulse: 620,
  puckShotMinHoldMs: 40,

  // puck debug
  puckDrawPickupRadius: false,
  puckDrawMagnetRadius: false,
  puckDrawState: false,
  puckDrawVelocity: false,

  // aim + crosshair
  controlScheme: 'WASD_MOVE_MOUSE_AIM',
  bodyFacingMode: 'MOVE_LAST',
  bodyTurnRate: 10.5,
  bodyTurnRateLowSpeedMult: 1,
  aimEnabled: true,
  aimMaxTurnRate: 10,
  aimDeadzonePx: 32,
  aimSmoothing: 0,
  stickAngleLimitEnabled: true,
  maxStickAngleFromBodyDeg: 75,
  stickAngleLimitSoftness: 1,
  snapEnabled: true,
  snapSpeedThreshold: 70,
  snapStrengthMax: 6,
  snapFadePower: 1.6,
  snapOnlyWhenInput: true,
  startLinearEnabled: true,
  startLinearOnThreshold: 41.1,
  startLinearOffThreshold: 75.35,
  startLinearSideKill: 10,
  startLinearAlignStrength: 12,
  startLinearRequiresInput: true,
  brakeAssistEnabled: true,
  brakeAssistDurationMs: 160,
  brakeAssistDragMult: 1.6,
  brakeMinSpeed: 40,
  alignSpeedThreshold: 80,
  alignStrength: 4,
  alignFadePower: 1.5,
  maxCurvatureEnabled: false,
  maxCurvatureLow: 11,
  maxCurvatureHigh: 2.6,
  lowSpeedAlignThreshold: 80,
  lowSpeedAlignStrength: 4,
  crosshairEnabled: true,
  crosshairSize: 16,
  crosshairThickness: 2,
  crosshairCenterGap: 4,
  hideSystemCursor: true,
  drawAimLine: false,
  drawTargetAngle: false,
  showTargetAngle: false,
  showHeading: false,
  drawVectors: false,
  showSnapFactor: false,
  showBrakeActive: false,
  showStartMode: false,
  drawVelComponents: false,
  drawMoveVector: true,
  drawBodyVector: true,
  drawAimVector: true,
  drawAimVectorRaw: false,
  drawAimVectorClamped: false,
  showAngles: false,
  showAngleDiff: false,
  debugDrawVectors: false,
  debugDrawArcPreview: false,

  __version: 1
};
